odoo/addons/base/models/res_users.py

```diff
@@ -1032,32 +1032,74 @@ def action_revoke_all_devices(self):
             'context': ctx,
         }
 
-    @api.model
-    @api.readonly
-    def has_group(self, group_ext_id):
-        # use singleton's id if called on a non-empty recordset, otherwise
-        # context uid
-        uid = self.id
-        if uid and uid != self._uid:
-            self = self.with_user(uid)
-        return self._has_group(group_ext_id)
+    def has_groups(self, group_spec: str) -> bool:
+        """ Return whether user ``self`` satisfies the given group restrictions
+        ``group_spec``, i.e., whether it is member of at least one of the groups,
+        and is not a member of any of the groups preceded by ``!``.
 
-    @api.model
-    @tools.ormcache('self._uid', 'group_ext_id')
-    def _has_group(self, group_ext_id):
-        """Checks whether user belongs to given group.
+        Note that the group ``"base.group_no_one"`` is only effective in debug
+        mode, just like method :meth:`~.has_group` does.
+
+        :param str group_spec: comma-separated list of fully-qualified group
+            external IDs, optionally preceded by ``!``.
+            Example:``"base.group_user,base.group_portal,!base.group_system"``.
+        """
+        if group_spec == '.':
+            return False
+
+        positives = []
+        negatives = []
+        for group_ext_id in group_spec.split(','):
+            group_ext_id = group_ext_id.strip()
+            if group_ext_id.startswith('!'):
+                negatives.append(group_ext_id[1:])
+            else:
+                positives.append(group_ext_id)
+
+        # for the sake of performance, check negatives first
+        if any(self.has_group(ext_id) for ext_id in negatives):
+            return False
+        if any(self.has_group(ext_id) for ext_id in positives):
+            return True
+        return not positives
+
+    def has_group(self, group_ext_id: str) -> bool:
+        """ Return whether user ``self`` belongs to the given group (given by its
+        fully-qualified external ID).
+
+        Note that the group ``"base.group_no_one"`` is only effective in debug
+        mode: the method returns ``True`` if the user belongs to the group and
+        the current request is in debug mode.
+        """
+        self.ensure_one()
+
+        if not (self.env.su or self == self.env.user or self._has_group('base.group_user')):
```

Comment:
Possibly `self._has_group('base.group_user')` should have been `self.env.user.has_group('base.group_user')`?

Regarding issue posted here: https://discord.com/channels/678381219515465750/1080100802338770964/1229802602879914034